add support for net40 and net45

moh-hassan opened the pull request:
- Add support for net40 and net45 for the netstandard2.0 project.
-  Resolve the null EntryAssembly Problem in issues:  #389  #424 forever.
- Pass the test case 'Add unit tests for Issue389 #[392]'(https://github.com/commandlineparser/commandline/pull/392)
- Create commandline multi-target project net40;net45;netcoreapp2.0 .
-  Ccreate commandline Test multi-target project net461;netcoreapp2.0 .
- Add class to support net40 
- Modify the cs demo to run in net40;net45;netcoreapp2.0;netcoreapp2.1
- Pass all tests (334 test case) in both net461 and netcoreapp2.0.
- Appveyor CI success on the fork [project](https://ci.appveyor.com/project/moh-hassan/commandline):

ericnewton76 commented:
Thanks for starting this.  It was next on my list.

Looks like fsharp project wont work on anything prior to 4.5.  We can probably split that build off for any <4.5.

I've been seeing requests for framework 3.5 support, although off the top of my head I can't think of how cumbersome that might be.

ericnewton76 commented:
This looks good, although the FSharp package is failing because FSharp isnt supported at the net40 level.

@moh-hassan  Any ideas on how to prevent it from building when doing a net40 build?  We can probably separate it out somehow.  The build is getting complicated... lol

moh-hassan commented:
Hi @ericnewton76 ,
You can use conditional build
Modify the xml project `CommandLine.csproj' and replace 

		  <ItemGroup>
			<PackageReference Include="FSharp.Core" Version="4.5.1" Condition="'$(BuildTarget)' == 'fsharp'" />
		  </ItemGroup>

With Condition:

		  
		  <ItemGroup Condition="'$(TargetFramework)' != 'net40' and '$(BuildTarget)' == 'fsharp'">
			<PackageReference Include="FSharp.Core" Version="4.5.1" Condition="'$(BuildTarget)' == 'fsharp'" />
		  </ItemGroup>

I  generated  Fsharp package by the command:

                 msbuild /t:pack /p:BuildTarget=fsharp

I get a package: `CommandLineParser.FSharp.2.5.0.nupkg' in my modified pR
the package include net40,net45,net461 and netstandard2

Let me know i this  done in your side, and ready for ping. I'm online now.

Can you check the generated package.
[CommandLineParser.FSharp.2.5.0.nupkg.zip](https://github.com/commandlineparser/commandline/files/2822703/CommandLineParser.FSharp.2.5.0.nupkg.zip)

ericnewton76 commented:
Gotcha.  I'll do that

Thanks again

On Fri, Feb 1, 2019 at 12:45 PM Mohamed Hassan <notifications@github.com>
wrote:

> Hi @ericnewton76 <https://github.com/ericnewton76> ,
> You can use conditional build
> Modify the xml project `CommandLine.csproj' and replace
>
> 	  <ItemGroup>
> 		<PackageReference Include="FSharp.Core" Version="4.5.1" Condition="'$(BuildTarget)' == 'fsharp'" />
> 	  </ItemGroup>
>
> With Condition:
>
> 	  <ItemGroup Condition="'$(TargetFramework)' != 'net40' and '$(BuildTarget)' == 'fsharp'">
> 		<PackageReference Include="FSharp.Core" Version="4.5.1" Condition="'$(BuildTarget)' == 'fsharp'" />
> 	  </ItemGroup>
>
> I generated Fsharp package by the command:
>
>              msbuild /t:pack /p:BuildTarget=fsharp
>
> I get a package: `CommandLineParser.FSharp.2.5.0.nupkg' in my modified pR
> the package include net40,net45,net461 and netstandard2
>
> Let me know i this done in your side, and ready for ping. I'm online now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/commandlineparser/commandline/pull/387#issuecomment-459806352>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACG_zKbCtBjizh4vVSPgfj9u6M0YRd-Xks5vJH1MgaJpZM4Z4uJU>
> .
>


-- 

E.Newton
Senior Software Craftsman
Cell: 407-497-1429
EMail: ericnewton76@gmail.com

moh-hassan commented:
@ericnewton76 
To check  that all tests pass in both net461 and netcoreapp2.0, I modified the Test Project as multi-target project:

                <TargetFrameworks>net461;netcoreapp2.0</TargetFrameworks>

I run `dotnet test` and find 14 test fail in net461 due to the reasons we discussed in the[ issue-389](https://github.com/commandlineparser/commandline/issues/389#issuecomment-455914655)

I fixed the failed test by using the solution I explained in the same issue

All tests pass Successfully in both net461 and netcoreapp2.0
The PR support: net40,net45,net461,netstandard2.0,netcoreapp2.0 and Fsharp fix is included.

The test pass in [CI Appveyor](https://ci.appveyor.com/project/moh-hassan/commandline)


with the green success text: 

**NET461**

>Test run for C:\projects\commandline\tests\CommandLine.Tests\bin\Debug\net461\CommandLine.Tests.dll(.NETFramework,Version=v4.6.1)
>....
>Total tests: 335. Passed: 335. Failed: 0. Skipped: 0.
>Test Run Successful.


**NETCoreApp2.0**
>Test run for C:\projects\commandline\tests\CommandLine.Tests\bin\Debug\netcoreapp2.0\CommandLine.Tests.dll(.NETCoreApp,Version=v2.0
>...
>Total tests: 335. Passed: 335. Failed: 0. Skipped: 0.
>Test Run Successful.

The confliction in the merge is:

		<<<<<<< master
			<TargetFrameworks>netstandard2.0;net40;net45;net461;netcoreapp2.0</TargetFrameworks>
		=======
			<TargetFrameworks>netstandard2.0; net461; netcoreapp2.0</TargetFrameworks>
		>>>>>>> master


which can  be resolved manually 
Let me know if  i missed something to do.

scriptator commented:
What's the status of this issue? It would be really great if it could be merged soon

vgalka-sl commented:
@moh-hassan @ericnewton76
Is there any progress on this? Where does it stand? May I offer my assistance if needed?
We really need the library to be compatible with .NET 4.5 for our product.

moh-hassan commented:
The PR is ready to be merged: 
It pass the unit test in netcoreapp2.0 and net461 and[ CI Appveyor](https://ci.appveyor.com/project
/moh-hassan/commandline)
It support: netstandard2.0;net40;net45;net461;netcoreapp2.0

The current Conflicting files:
src/CommandLine/CommandLine.csproj 

>			  <<<<<<< master	
>        <TargetFrameworks>netstandard2.0;net40;net45;net461;netcoreapp2.0</TargetFrameworks>
>			=======
>				<TargetFrameworks>netstandard2.0; net461; netcoreapp2.0</TargetFrameworks>
>			>>>>>>> master

**It Can be  resolved to:**

		<TargetFrameworks>netstandard2.0;net40;net45;net461;netcoreapp2.0</TargetFrameworks>

tests/CommandLine.Tests/Unit/Text/HelpTextTests.cs

>		<<<<<<< master
>			public class HelpTextTests: BaseTest
>		=======
>			public class HelpTextTests : IDisposable
>		>>>>>>> master

can be  resolved to:

	public class HelpTextTests: BaseTest,IDisposable 

@ericnewton76 
Is there any modification that I can do for merging the PR

moh-hassan commented:
@vgalka-sl 
The PR is ready to be merged after resolving the confliction as I [explained above](https://github.com/commandlineparser/commandline/pull/387#issuecomment-479605350).

vgalka-sl commented:
> @vgalka-sl
> The PR is ready to be merged after resolving the confliction as I [explained above](https://github.com/commandlineparser/commandline/pull/387#issuecomment-479605350).

@moh-hassan Thank you! Seems Eric is busy, so we just have to wait.
Btw, maybe you can rebase and resolve the conflict yourself to make it "cleaner"?

moh-hassan commented:
@ericnewton76 
@vgalka-sl 

I have to resolve an error (not related to the current PR) after rebasing , may be due to merging a previous commit: 

>A compilation error: 
>Error	CS0102	The type 'ErrorType' already contains a definition for 'VersionRequestedError'

'VersionRequestedError' is duplicated in the same file: Error.cs in [line63 ](https://github.com/commandlineparser/commandline/blob/675d4b75c714267938bd9efdb0d160a23bfeae22/src/CommandLine/Error.cs#L63)and [line68](https://github.com/commandlineparser/commandline/blob/675d4b75c714267938bd9efdb0d160a23bfeae22/src/CommandLine/Error.cs#L68)

Also, resolving the following compilation errors:
- CS0160	A previous catch clause already catches all exceptions of this or of a super type ('Exception')	CommandLine (commandline\src\CommandLine\Core\ReflectionExtensions.cs	113)
- Error	CS0103	The name 'InvalidAttributeConfigurationError' does not exist in the current context	CommandLine (src\CommandLine\Core\ReflectionExtensions.cs	115)

Can you confirm these modifications?

**Edit:**
There is a compilation error of the last merge. Have a look to this [discussion](https://github.com/commandlineparser/commandline/pull/314#issuecomment-480961683).

moh-hassan commented:
The new [PR#428 ](https://github.com/commandlineparser/commandline/pull/428)submitted by @Wind010   is fine and fix the issue of [compilation errors ](https://github.com/commandlineparser/commandline/pull/387#issuecomment-480876862) and I can rebase Upstream after it's (PR#428) merged.

moh-hassan commented:
Moving the PR to a separate branch in a new [PR# 430](https://github.com/commandlineparser/commandline/pull/430)
